### ISSUE TITLE
Evalaute Profile-Guided Optimization (PGO)

### ISSUE THREAD

**Issue by zamazan4ik** [meilisearch/meilisearch#3958]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are [here](https://github.com/zamazan4ik/awesome-pgo/).

Since PGO showed measurable improvements in many loads (including many databases) I think it could be useful to check PGO on Meilisearch.

We need to perform PGO benchmarks on Meilisearch. And if it shows improvements - add a note about possible improvements Meilisearch performance with PGO. Providing an easier way (e.g. a build option) to build scripts with PGO can be useful for the end-users too. Another option of how to use Meilisearch is to optimize provided cloud installation.

For the Rust projects, I recommend starting with https://github.com/Kobzol/cargo-pgo .

P.S. Not sure if is it right for such an issue. If you know a better place for the issue (maybe Discussions?) - feel free to convert it to the right type.

**Comment by zamazan4ik** [comment]
Yesterday I did Profile-Guided Optimization of Meilisearch and want to share my results.

## Test environment

* Fedora 38
* Linux kernel 6.4.6 (default Linux kernel from Fedora repos)
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Rustc: 1.71 (but didn't check does Meilisearch overwrite Rustc version or not)
* Meilisearch: latest `main` commit at the moment (`f391039a6f63d8414f0b1437e79ca9f3dcc0cf2a`)

## Bench

As a benchmark, I use *all* built-in Meilisearch benchmarks from the `benchmarks` directory. The benchmarks are used as a training and evaluation set. The optimization was done via [cargo-pgo](https://github.com/Kobzol/cargo-pgo). I compared Release (a first bench run) vs PGO (a second bench run). So you need to interpret the results like "How much PGO mode is faster/slower than Release mode".

The results are available here: https://pastebin.com/WYzMMVPP

The results right now are quite disappointing. On the one hand, indexing performance was improved by ~5% on average - not bad. But the search performance was degraded much worse - sometimes by 3 times.

I thought it could be mitigated by training on the search suite with `cargo pgo bench -- --bench search_songs` but got the same results. We need to investigate further why the search performance was affected so much by PGO. My guess right now is that during the `search_songs` somewhere there is a load for indexing and it affects the search performance. Another guess - just a "bug" in the PGO implementation so the compiler somewhere makes a wrong optimization decision.

**Comment by irevoire** [comment]
> My guess right now is that during the search_songs somewhere there is a load for indexing and it affects the search performance.

Yes, that’s the case. We index the `smol_songs` dataset once at the very beginning and then run all the search benchmarks on this index.

**Comment by Kerollmops** [comment]
Hey @zamazan4ik 👋

Thank you very much for the time spent doing these PGO experimentations. I am happy to see the indexation part improved by 5% _FOR FREE_ 🎁

However, a lot of time has passed since then, and [we improved the search algorithm in v1.2.0](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0), and we have seen enormous impacts on the search performance 🚀 Do think you could have the time to make a new try with the latest version of Meilisearch.

Is it possible to expose an experimental binary with PGO enabled on our CI?

Have a nice day, and sorry again for the late reply 🍕

**Comment by zamazan4ik** [comment]
> However, a lot of time has passed since then, and [we improved the search algorithm in v1.2.0](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0), and we have seen enormous impacts on the search performance 🚀 Do think you could have the time to make a new try with the latest version of Meilisearch.

Yup, I could try to repeat my PGO benchmarks. However, I think the performance impact should be the same since PGO does not rely on algorithmic optimizations or something like that. PGO "magic" usually is just a much better inlining decisions and similar stuff.

> Is it possible to expose an experimental binary with PGO enabled on our CI?

Yes, it's possible to do. You just need to build an additional `meilisearch` version with PGO optimization. The main concern here - is how to collect a proper working profile. As you see above, using the Meilisearch benchmarks could lead to even worse results. We need to evaluate other ways of collecting proper profiles.

**Comment by Kerollmops** [comment]
> However, I think the performance impact should be the same since PGO does not rely on algorithmic optimizations or something like that.

We modified the whole search algorithm so there are great chance that the PGO optimization results would be different. However, you are absolutely right that we must find a better way to collect profiles with better datasets and use cases!
